File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    int[] dy = {0, 1, 0, -1};
    int[] dx = {1, 0, -1, 0};
    public int[][] solution(int n) {
        int[][] answer = new int[n][n];
        int y = 0, x = 0;
        int dir = 0;
        int start = 1;
        while(start <= n * n) {
            answer[y][x] = start++;
            int ny = y + dy[dir];
            int nx = x + dx[dir];
            if(ny < 0 || ny >= n || nx < 0 || nx >= n || answer[ny][nx] != 0) {
                dir = (dir + 1) % 4;
                ny = y + dy[dir];
                nx = x + dx[dir];
            }
            y = ny;
            x = nx;
        }
        
        return answer;
    }
}

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    answer = [n * [0] for _ in range(n)]
    # print(answer)
    
    val = 1
    y, x = 0, 0
    dy = [0, 1, 0, -1]
    dx = [1, 0, -1, 0]
    dir = 0
    while val <= n * n:
        answer[y][x] = val
        val += 1
        ny = y + dy[dir]
        nx = x + dx[dir]
        if ny < 0 or ny >= n or nx < 0 or nx >= n or answer[ny][nx] != 0:
            dir = (dir + 1) % 4
            ny = y + dy[dir]
            nx = x + dx[dir]
        y = ny
        x = nx
        
    return answer